protocols/gossipsub/src/handler.rs

```diff
@@ -382,6 +321,77 @@ impl EnabledHandler {
             }
         }
 
+        // If the outbound stream was not established then the send queue remains unprocessed.
+        // We put inbound stream processing on hold if the queue is full to avoid overloading
+        // it with new response messages from the behaviour.
+        if self.send_queue.len() >= SEND_QUEUE_CAPACITY {
+            return Poll::Pending;
```

Comment:
Thinking out loud here: We must ensure that a waker is registered when we return `Poll::Pending` here.

The above `loop` only breaks in two scenarios:

- An outbound stream operation returned `Poll::Pending`
- We don't have an outbound stream

The former case registers a waker. The latter does **not**. thus, I think we should put an additional guard here that checks if `outbound_stream.is_some()` otherwise we are violating the "poll"-rule.

I am somewhat tempted to say: Let's land this in two separate PRs:

1. Change the order in which we prioritize the streams. This is totally safe and should yield improvements because we are emptying local buffers first.
2. Consider bailing early here if the outbound stream can't make progress. I am not super familiar with gossipsub but holistically, this might have implications on the network if we stop reading from a stream because the other party stopped reading from our stream (or is slow in doing so). Perhaps that is a good thing though?

Reply:
> The former case registers a waker. The latter does not. thus, I think we should put an additional guard here that checks if outbound_stream.is_some() otherwise we are violating the "poll"-rule.

Could we just always schedule a wake-up with `cx.waker().wake_by_ref()` if `outbound_stream.is_none()` to uphold the poll contract?

> I am somewhat tempted to say: Let's land this in two separate PRs:
Sure! I can do that instead since it's a straightforward change.

> Consider bailing early here if the outbound stream can't make progress. I am not super familiar with gossipsub but holistically, this might have implications on the network if we stop reading from a stream because the other party stopped reading from our stream (or is slow in doing so). Perhaps that is a good thing though?

My reasoning was that this establishes some backpressure to the remote and stops it from sending more messages, or at the minimum prevents the send-queue from growing (potentially) indefinitely.  Imo it might be worth it to slow down ingestion to match the outbound processing rate.

Reply:
> > The former case registers a waker. The latter does not. thus, I think we should put an additional guard here that checks if outbound_stream.is_some() otherwise we are violating the "poll"-rule.
> 
> Could we just always schedule a wake-up with `cx.waker().wake_by_ref()` if `outbound_stream.is_none()` to uphold the poll contract?

Wouldn't that lead to an endless loop of continuous wakeups because we don't have any other task before the early return that will return `Poll::Pending`?